### ISSUE TITLE
if an exceptions raised when Pipfile is about to be modified, will render Pipfile empty

### ISSUE THREAD

**Issue by chenull** [kennethreitz/pipenv#225]
In experimenting `pipenv` with more advanced `pip` arguments, e.g install from `git+http://repo/some.git` or `git+file:///Users/ayik/Repo/django`, if there is an exceptions, will render `Pipfile` empty (wiped)

```
$ pipenv --version
pipenv, version 3.4.1`
```

### Install from git+http

`$ pipenv install git+https://github.com/django/django.git`

```shell
Installing git+https://github.com/django/django.git...
Collecting git+https://github.com/django/django.git
  Cloning https://github.com/django/django.git to /private/var/folders/t_/cpyg93cx79zfjwgnp24fx7rw0000gn/T/pip-dfds2x7k-build
Requirement already satisfied: pytz in /Users/ayik/.virtualenvs/empu/lib/python3.6/site-packages/pytz-2016.10-py3.6.egg (from Django==2.0.dev20170214233504)
Installing collected packages: Django
  Found existing installation: Django 1.10.5
    Uninstalling Django-1.10.5:
      Successfully uninstalled Django-1.10.5
  Running setup.py install for Django: started
    Running setup.py install for Django: finished with status 'done'
Successfully installed Django-2.0.dev20170214233504

Adding git+https://github.com/django/django.git to Pipfile's [packages]...
Traceback (most recent call last):
  File "/Users/ayik/.local/bin/pipenv", line 11, in <module>
    sys.exit(cli())
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/cli.py", line 771, in install
    project.add_package_to_pipfile(package_name, dev)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/project.py", line 173, in add_package_to_pipfile
    self.write_toml(p)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/project.py", line 127, in write_toml
    f.write(format_toml(toml.dumps(data)))
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/toml.py", line 602, in dumps
    addtoretval, addtosections = _dump_sections(sections[section], section)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/toml.py", line 619, in _dump_sections
    if not re.match(r'^[A-Za-z0-9_-]+$', section):
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/re.py", line 172, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
```

### Install from local repository:

`pipenv install 'git+file:///Users/ayik/Repo/Django/django`

```shell
Installing git+file:///Users/ayik/Repo/Django/django...
Collecting git+file:///Users/ayik/Repo/Django/django
  Cloning file:///Users/ayik/Repo/Django/django to /private/var/folders/t_/cpyg93cx79zfjwgnp24fx7rw0000gn/T/pip-aqyyogjv-build
Requirement already satisfied: pytz in /Users/ayik/.virtualenvs/empu/lib/python3.6/site-packages/pytz-2016.10-py3.6.egg (from Django==2.0.dev20170214005208)
Installing collected packages: Django
  Running setup.py install for Django: started
    Running setup.py install for Django: finished with status 'done'
Successfully installed Django-2.0.dev20170214005208

Adding git+file:///Users/ayik/Repo/Django/django#editable=1 to Pipfile's [packages]...
Traceback (most recent call last):
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/packaging/requirements.py", line 90, in __init__
    req = REQUIREMENT.parseString(requirement_string)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 1617, in parseString
    raise exc
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 1607, in parseString
    loc, tokens = self._parse( instring, 0 )
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 3376, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 1383, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pyparsing.py", line 3164, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected stringEnd (at char 3), (line:1, col:4)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2890, in __init__
    super(Requirement, self).__init__(requirement_string)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/packaging/requirements.py", line 94, in __init__
    requirement_string[e.loc:e.loc + 8]))
packaging.requirements.InvalidRequirement: Invalid requirement, parse error at "'+file://'"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ayik/.local/bin/pipenv", line 11, in <module>
    sys.exit(cli())
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/cli.py", line 771, in install
    project.add_package_to_pipfile(package_name, dev)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/project.py", line 166, in add_package_to_pipfile
    package = convert_deps_from_pip(package_name)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/utils.py", line 31, in convert_deps_from_pip
    req = [r for r in requirements.parse(dep)][0]
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pipenv/utils.py", line 31, in <listcomp>
    req = [r for r in requirements.parse(dep)][0]
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/requirements/parser.py", line 50, in parse
    yield Requirement.parse(line)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/requirements/requirement.py", line 181, in parse
    return cls.parse_line(line)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/requirements/requirement.py", line 159, in parse_line
    pkg_req = Req.parse(line)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2935, in parse
    req, = parse_requirements(s)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2883, in parse_requirements
    yield Requirement(line)
  File "/Users/ayik/.local/venvs/pipenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2892, in __init__
    raise RequirementParseError(str(e))
pkg_resources.RequirementParseError: Invalid requirement, parse error at "'+file://'"
```

and... my `Pipfile` is now an empty file ! 😱

**Comment by nateprewitt** [comment]
Thanks for opening this @chenull, this is definitely not what we want happening here. I've added be68acc which will prevent the accidental blank Pipfile when an exception is raised.

As for handling the install, there's a couple issues here. The first one being, due to our upstream `pkg_resources` we currently can only properly parse vcs dependencies that have a `#egg=ProjectName` at the end. This could probably be documented better, and it may be worth considering raising an exception if we can't find it before installation.

The second issue has to do with the use of the file:// scheme on the vcs URI. This also doesn't appear to be supported in `pkg_resources`. I haven't taken a look there to see if either of these have an open issue but we can see if there's work that can be done there.

In the meantime, I think we'll have to simply suggest specifying the `#egg=ProjectName` approach with version control systems we support (git, mecurial, svn, and bazaar).

**Comment by nateprewitt** [comment]
The file erasure is resolved in v3.4.2 which I just released. The git issues are a longer running problem that likely won't be addressed until 3.5.0 at the earliest but possibly not until a 4.0.0 release. Thanks again for bringing this to our attention @chenull!

Please refer to ticket #204 for the longer running git discussion.
